Byte counting at Congestion avoidance (#3917)
quicwg/base-drafts: draft-ietf-quic-recovery.md

  the mode is slow start and the window grows by the number of bytes
  acknowledged.

bytes_acked:

> : Records number of bytes acked, and used for incrementing

Reviewer: ```suggestion
:  Number of bytes acknowledged which have not yet been used to increment
```

> : Records number of bytes acked and used during congestion
>   avoidance, for incrementing the congestion window.

Reviewer 2: ```suggestion
: Records number of bytes acked, and used for incrementing
  the congestion window during congestion avoidance.
```

  the mode is slow start and the window grows by the number of bytes
  acknowledged.

bytes_acked:

> : Store how many bytes acked so far. During congestion avoidance,
>   congestion window will be incremented by max_datagram_size
>   when this value is same or bigger than the congestion window.

Reviewer 2: ```suggestion
: Records number of bytes acked and used during congestion
  avoidance, for incrementing the congestion window.
```

Author: LGTM. Updated.

      congestion_window += acked_packet.sent_bytes
    else:
      // Congestion avoidance.
      bytes_acked += acked_packet.sent_bytes
      if (bytes_acked >= congestion_window):
        bytes_acked -= congestion_window

>         congestion_window += max_datagram_size

Reviewer: This part contains the larger change.  Previously, congestion_window was adjusted in 1 byte increments.  Now it's adjusted in max_datagram_size increments.  In practice, it likely doesn't matter that much, but I am concerned about perpetuating packet-size quantization in congestion control, since my impression is that it's an artifact of the Linux kernel, not an intentional design choice.

I believe BSD-based OS's use bytes instead of packets, and bytes always seemed more correct to me.

However, this isn't that difficult to fix.  How about:
  cwnd_in_packets = congestion_window / max_datagram_size
  congestion_window += bytes_acked / cwnd_in_packets
  bytes_acked = bytes_acked % cwnd_in_packets

Reviewer 2: I don't see how this is packet-based computation. Yes, the cwnd is incremented only when it can be incremented in max_datagram_size chunks, but at any given time, you know exactly what your fractional component is: (bytes_acked/congestion_window). At any given point in time, your total send allowance, in bytes, is (congestion_window + bytes_acked * max_datagram_size / congestion_window).

This fractional difference only matters if you are sending packets smaller than max_datagram_size.

Author: FYI, FreeBSD uses byte counting by default and use the same mechanism of this PR: https://github.com/freebsd/freebsd/blob/0b7e58eb670d67cdc0268be5d588839038aa5abf/sys/netinet/tcp_input.c#L355

And I think @ianswett 's code has a problem
> bytes_acked = bytes_acked % cwnd_in_packets
this will make bytes_acked too small (only between 0 - cwnd_in_packets - 1 range)

I'm happy with the current logic.

Reviewer 3: > This fractional difference only matters if you are sending packets smaller than max_datagram_size.

This isn't quite right.  Based on this, you need to receive a complete window before you increase the window size.  Previously, you could maybe increment by smaller amounts sooner.  The difference is between a continuous increase (or at least an increase with every acknowledgments) and a stepped increase.  Let's say max_datagram_size is 1000 and your window is currently 10,000, you will have to receive acknowledgements for 10 packets before you can go to 11,000.  If you increase progressively, then you might go to 10,100 then 10,200, and so forth sooner.

If the assertion is that you can't use that small extra amount, consider that the window could be 9,999 to begin with.  That is entirely likely as you rarely settle on a strict multiple of packet size.  The first arriving acknowledgment would allow another full-sized packet to be sent much sooner than otherwise.

So this is a functional change, even if it you might easily decide that it is acceptable.

Reviewer 3: Oh, I should probably concede that more increases to the congestion window will result in the final congestion window being smaller, so you are trading a larger final increase (this PR) with earlier availability of bytes (status quo).

Say you get 10 acks for 1000 bytes, with a starting cwnd of 10,000 and max_datagram_sizeo of 1000.

With this code, you end up with no change until the last ack, where you go to 11,000.

With the previous code, you would go to: 10100, 10199, 10297, 10394, 10490, 10585, 10680, 10773, 10866, 10958.  (That's using double-precision logic, you lose two bytes if you round.)

That's less in the end, but the value of that to you depends on whether you can use any of those intermediate values.

Reviewer 3: If you want to realize increments immediately, but not lose any credit, then I think that the algorithm is:

```
credit = saved_credit + (bytes_acked * max_datagram_size)
congestion_window += credit / congestion_window
saved_credit = credit % congestion_window
```

This uses integer math to produce a value that matches the double-precision logic from my previous comment.

Reviewer: Yesterday was a long day, so I suspect my pseudocode has an error, but as @mt points out this does change how the pseudocode works.

Author: In the scenario @martinthomson mentioned, I think the problem with small increase is that it can result in a small sized packets.

e.g. bytes_in_flight = 10000, ssthresh = cwnd = 10000, mss = 1000
  -> received ack of 1000 bytes
  -> byts_in_flight = 9000, cwnd = 10100

Now sender can send 1100 bytes, can result in sending 2 packets, with 1000 bytes and 100 bytes each when there is a enough size of data is ready. The pseudocode in this PR doesn't have such issue.

Author: Briefly tested using quiche (https://github.com/cloudflare/quiche/pull/610) -- when downloading 10Mbytes file using 10Mbit/40ms link, total # of sender packets can be reduced by 14% (mostly small packets than MSS) with slightly better performance.

Reviewer: The small fragmented packet issue is an interesting one, is there an existing RFC which discusses this we could cite with advice on how to avoid overly small packets?

Reviewer 4: What about RFC 3465?

Author: This mechanism described in RFC3465 as in the original PR description but the RFC itself has no pseudocode. I don't know other references discussing about those small fragments due to byte counting in congestion avoidance -- probably it's because most of TCP RFC use window-based cwnd, not byte based?

Reviewer 2: Yes, this is a change in how the sender works, but it's not a significant difference. And it is in the pseudo code in the appendix, not in the text. 

Let's be clear that we are defining something that we are calling Reno for QUIC, knowing fully well that it's just one of many controllers that people will attempt. The current pseudo code is not a reference, it is one way of doing Reno. What this PR proposes is what a number of implementations already do.

I don't think there's much more to see here. I'm fine with not including the PR, or with including the PR. Most implementations are likely to implement it the way the PR does, but the current pseudo code is also good in that it is algorithmically clear.

Reviewer: I think this PR is fine, but I'm also not seeing a big reason for changing the current pseudocode.

It turns out Chrome doesn't do it this way anymore, but we used to.

Author: I implemented ABC for cubic in quiche https://github.com/cloudflare/quiche/pull/624 and did some evaluation:
Measured using quiche apps (quiche-server and quiche-client) in 5mbps/60ms rtt/0% loss. Also last highlighted row is when I changed MAX_DATAGRAM_SIZE to 1350 to match with what quiche apps use (quiche has MAX_DATAGRAM_SIZE constant). 

![Screen Shot 2020-08-28 at 5 21 49 PM](https://user-images.githubusercontent.com/1229714/91624295-43120300-e954-11ea-8559-9028131dbfdb.png)

You can see it can save some amount of sent packets (3.8% for cubic, 6.7% for reno, 9.3% for cubic and 1350 mss aligned) with slightly faster download time.

I think it's up to editors for deciding to merge this PR not, but I believe it's good for implementers to start with less issue.

Reviewer 2: I'm surprised to see these differences, @junhochoi -- why are the lost packets so different? Why's there any loss when the network has 0% loss?

Author: @janaiyengar it's basically loss by congestion. e.g. https://blog-cloudflare-com-assets.storage.googleapis.com/2020/05/cubic-hs.png (cubic+hystart case). All values are median number over 10 run. And lost packet here is the number of lost packets detected by quiche sender, so actual packet loss may be lower than that.

Reviewer 2: Are these controlled at the bottleneck or is this uncontrolled loss on the Internet? The number of packets lost seem quite different, and I'm trying to understand why those loss numbers are so different.

Author: This is a simulated environment, not an public internet. In this testing I used MacOS dummynet. As I said this number is what recovery module detected as loss (which will trigger a congestion event). So may not be an actual loss (e.g. delivered late after a time threshold). I didn't measure actual packet loss.
What's your expectation?

      congestion_window += acked_packet.sent_bytes
    else:
      // Congestion avoidance.
      bytes_acked += acked_packet.sent_bytes

>       if (bytes_acked >= congestion_window):
>         bytes_acked -= congestion_window
>         congestion_window += max_datagram_size

Reviewer 2: There's a minor behavior difference here from the old code, in that the cwnd here increases by at most 1. I think that's fine; I don't think it's possible to have an increase that is larger. The following code is the same as the old behavior, but I don't think it's necessary.

```suggestion
      while (bytes_acked >= congestion_window):
        bytes_acked -= congestion_window
        congestion_window += max_datagram_size
```

Author: I don't think `while()` is necessary because `OnPacketAcked()` (recently changed) is called per acked packet, so `acked_packet.sent_bytes` is always smaller than `max_datagram_size`.

Reviewer 4: I agree with @junhochoi.

Reviewer 2: Agreed. I wanted to note the difference, but as I said, I don't think it's necessary.

     if (!InCongestionRecovery(sent_time)):
       congestion_recovery_start_time = now()
       congestion_window *= kLossReductionFactor

>        bytes_acked *= kLossReductionFactor

Reviewer: I thought this would be set to 0 here.  Why reduce it?

Reviewer 2: The current cwnd is (cwnd + bytes_acked/cwnd). Multiplying that by kLossReductionFactor means you multiply bytes_acked by kLossReductionFactor as well.

Author: Just thinking again... this is when a recovery episode starts and we enter congestion avoidance by ssthresh=cwnd. And bytes_acked is only used in congestion avoidance and doing nothing during slow start.

I think we simply clear bytes_acked to 0 as @ianswett said because it's a new start of congestion avoidance, or no need to touch bytes_acked because even if there is a previous leftover bytes_acked, it will not impact much.

Reviewer: Thanks for the explanation.  I think I'm fine either way, but questions like this are one of the reason I liked the old approach of assuming it's the implementation's responsibility to deal with rounding/etc.

Reading this again, I realize there's a somewhat substantive difference in behavior between the previous pseudocode and this PR, because the previous pseudocode would increment CWND in bytes, not only max_datagram_size.

Reviewer 2: This isn't a big difference, so I'm fine with any resolution here. @ianswett : I don't understand what you mean by the substantive difference. The previous code incremented by (bytes_acked/cwnd) fraction of max_datagram_size on each ack, this one accumulates that fraction until it is 1 and then adds max_datagram_size. It is functionally the same.

Reviewer: I agree that this is a small difference either way.

       congestion_window *= kLossReductionFactor
       congestion_window = max(congestion_window, kMinimumWindow)
       ssthresh = congestion_window

>        bytes_acked *= kLossReductionFactor

Reviewer 2: move this up to the line below where the congestion window is reduced

Author: Done. Thanks.
